operators/pkg/controller/apmserver/apmserver_controller_test.go

```diff
@@ -0,0 +1,379 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package apmserver
+
+import (
+	"testing"
+
+	apmv1alpha1 "github.com/elastic/cloud-on-k8s/operators/pkg/apis/apm/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/certificates"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/certificates/http"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/keystore"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/watches"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/utils/k8s"
+	"github.com/go-test/deep"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+var certSecretName = "test-apm-server-apm-http-certs-internal"
+
+type testParams = DeploymentParams
+
+func (tp testParams) withConfigChecksum(checksum string) testParams {
+	tp.PodTemplateSpec.Labels["apm.k8s.elastic.co/config-file-checksum"] = checksum
+	return tp
+}
+
+func (tp testParams) withVolume(i int, volume corev1.Volume) testParams {
+	tp.PodTemplateSpec.Spec.Volumes = append(tp.PodTemplateSpec.Spec.Volumes[:i], append([]corev1.Volume{volume}, tp.PodTemplateSpec.Spec.Volumes[i:]...)...)
+	return tp
+}
+
+func (tp testParams) withVolumeMount(i int, mnt corev1.VolumeMount) testParams {
+	for i := range tp.PodTemplateSpec.Spec.Containers {
+		existing := tp.PodTemplateSpec.Spec.Containers[i].VolumeMounts
+		tp.PodTemplateSpec.Spec.Containers[i].VolumeMounts = append(existing[:i], append([]corev1.VolumeMount{mnt}, existing[i:]...)...)
```

Comment:
I am confused by this line.
a)  `i` is the index of the container being inspected but then the same index is used to access the volume mounts array of that container. Will the size of the volume mounts array always be greater than `i`? If not there will be panic.
b) Is it important to insert the new mount in the middle?

Reply:
Well spotted, I wonder why this worked at all. 

> b) Is it important to insert the new mount in the middle?

Yes and no. The code that we are testing here inserts volume mounts at different places unfortunately, because it happens in different functions. I did not want to refactor this right now, given that this is supposed to go into 0.9.